common/src/main/java/com/box/l10n/mojito/okapi/filters/Md5ContextAwareIdGenerator.java

```diff
@@ -0,0 +1,55 @@
+package com.box.l10n.mojito.okapi.filters;
+
+import com.google.common.collect.HashMultiset;
+import org.apache.commons.codec.digest.DigestUtils;
+
+/**
+ * An id generator that uses md5 of the current text unit content, the previous content and a
+ * sequence to generate near stable ids.
+ *
+ * <p>The goal of this generator is to generate stable ids that points to the same source content in
+ * between document revisions. Specifically, when the document structure only is modified or small
+ * source edits are made, it avoids generating a brandnew set of text units that will require full
+ * re-processing.
+ *
+ * <p>For example the default HTML {@link net.sf.okapi.common.IdGenerator} uses sequences. If using
+ * those ids as text unit names then any change to the HTML structure leads to all the text unit
+ * names to point to different source content between different revisions.
+ *
+ * <p>{@code <p>a</p><p>b</p>} will generate the following pairs {@code tu1: a, tu2: b} and {@code
+ * <p>n</p><p>a</p><p>b</p>} will generate the following pairs {@code tu1: n, tu2: a, tu3: b}
+ *
+ * <p>Instead, Md5ContextAwareIdGenerator generates an id based on the current source, the previous
+ * source and the number of occurence of the previous source. The number of occurrence is used to
```

Comment:
I was thinking a bit more about the algorithm, what does the hash of the previous source give us?

Wouldn't the hash of the current source + the current position/number of the occurrences of the source text in the document be enough - and also result in less cascading behaviour (less impact to the items following the duplicate unit changes)?

Reply:
@garionpin any thoughts on this?

Reply:
Very exhaustive documentation and options, thank you for setting that up!

As I went through the different ID generators - it occurred to me that the "default" generator type for the majority of (extracted) content that Mojito processes currently doesn't resemble any of them - and is actually a simple "hash" generator - where we record usages and do not differentiate between the different instances of the same string - and thus do not have the ability to translate them differently.

Starting with that in mind - what is the ideal text unit data modelling and processing ability that we need/want in Mojito?
- Should we optimize for reusing translations/reducing translation costs without the ability to disambiguate translations for identical instances? This has the advantage of simplicity but requires all translations to be the same for all identical instances.
- Should we support out of the box different instances of the same identical string to have different translations ("forked" translations)? (Where until that is needed - they only have a single entity in the system).

I would note that having duplicate entries with the same values should be assumed to incur extra costs by default.

A certain subset of TMSs might treat them as the "same" string when translating from scratch - if certain settings are configured (E.g.: Smartling) - though that is not something that can necessarily be assumed by default for third party TMSs in general.

Either way, we should answer the above question to ensure we have a target goal for consistency for different file types (and selecting the right ID generator).

Reply:
> Very exhaustive documentation and options, thank you for setting that up!
> 
> As I went through the different ID generators - it occurred to me that the "default" generator type for the majority of (extracted) content that Mojito processes currently doesn't resemble any of them - and is actually a simple "hash" generator - where we record usages and do not differentiate between the different instances of the same string - and thus do not have the ability to translate them differently.
> 

I'm not sure to fully understand which steps you are referring to: pre mojito string extraction, mojito filter, mojito text unit creation, etc. 

Adding some thoughts...

Most formats processed by Mojito are resource bundles. They come with a concept of ID defined by the format itself. Source strings can be easily duplicated and have different translations as long as the IDs are different. 

For the PO files, Mojito behaves a bit differently from vanilla Okapi but maps to the Gettext semantics to replicate the unique id (msg+context) concept the library has. The "location" element is a meta data that does not differentiate the string at runtime. As such it should not be a criteria for having different translations either. 

Some file formats like "properties files" can have duplicate entries "id=source". Java runtime won't differentiate them, nor fail loading them but it might override the "first value" with the "last value" for the given ID. In that case, Mojito considers them to be a single text unit ([skip duplicate logic in asset extraction](https://github.com/pinterest/mojito/blob/f070b7ea45ccc53d866928785256b50c9aac4b6d/common/src/main/java/com/box/l10n/mojito/okapi/extractor/AssetExtractionStep.java#L76)) but the localized files will have both entries translated to make sure both "keep first'' and "keep last" logic will work at runtime.

For random HTML files, typically there are no ids. Adding this generator is an attempt at supporting minor changes between versions of the asset without re-creating a full text unit set. The generator could be a filter option and support vanilla Okapi ids, and then just rely on a more standard leveraging workflow.

> Starting with that in mind - what is the ideal text unit data modelling and processing ability that we need/want in Mojito?
> 
> * Should we optimize for reusing translations/reducing translation costs without the ability to disambiguate translations for identical instances? This has the advantage of simplicity but requires all translations to be the same for all identical instances.

I guess the definition of "identical instances" here is the key. I'm not sure we have the same thing in mind.

"almost identical", i.e. duplicated source without id (like html) or duplicated source with different ids (resource bundle) must support different translations. 

"truly identical", i.e. duplicated source with id (buggy properties files) don't need different translations. 

Current system skips “truly identical” and supports “almost identical”. I think that is a must have but maybe you have something different in mind when you say “identical”.

The cost saving aspect is secondary to that design choice in my opinion since it can be handled by systematically leveraging by source and it would provide similar capabilities (need to make sure translations stay in sync, but that is doable too). Trying to have near stable IDs is also about having a more clear history and quicker TAT, but avoid recreating new text units all the time. 

> * Should we support out of the box different instances of the same identical string to have different translations ("forked" translations)? (Where until that is needed - they only have a single entity in the system).

I think there are more advantages to have a simple "asset to list of text unit mapping" and then translation linked to the text units directly. Then leveraging can do the work of associating different text units together and copy translations over. 

I don't see a real benefit of having a data model organized around "source hash" with a default translation and having overrides based on context" when necessary.

> 
> I would note that having duplicate entries with the same values should be assumed to incur extra costs by default.
> 

that can be easily mitigated with leveraging, no? 

> A certain subset of TMSs might treat them as the "same" string when translating from scratch - if certain settings are configured (E.g.: Smartling) - though that is not something that can necessarily be assumed by default for third party TMSs in general.
> 

IMHO it is not reasonable design choice (and not a good default setting either). If it was we would not bother with ID in the code in a first place. 

> Either way, we should answer the above question to ensure we have a target goal for consistency for different file types (and selecting the right ID generator).

It is the first time we add support for a format that no standard id, or extractable id. That id generator tries to fill some gaps by leveraging might be a good alternative.